### PR TITLE
Introducing Deterministic  Configuration to WasiCtxBuilder

### PR DESCRIPTION
**Pull request by cameronbarker**
A new approach based on #268.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -17,4 +17,5 @@ end
 
 group :test do
   gem "rspec", "~> 3.1"
+  gem "pry"
```
Comment (cameronbarker) [inline Gemfile:20]: Need to remove.

File: Gemfile.lock [hunk 2]
```diff
@@ -100,4 +106,4 @@ DEPENDENCIES
   yard-rustdoc (~> 0.3.2)
 
 BUNDLED WITH
-   2.5.4
+   2.4.18
```
Comment (cameronbarker) [inline Gemfile.lock:109]: Need to fix

File: ext/src/ruby_api/wasi_ctx_builder.rs [hunk 3]
```diff
@@ -192,38 +195,56 @@ impl WasiCtxBuilder {
         rb_self
     }
 
+    pub fn set_deterministic(rb_self: RbSelf) -> RbSelf {
```
Comment (cameronbarker) [inline ext/src/ruby_api/wasi_ctx_builder.rs:198]: I felt that this name might be incorrect given that it doesn't take a parameter. I had though potentially `as_deterministic`, but would appreciate any recommendations.

File: ext/src/ruby_api/wasi_ctx_builder.rs [hunk 4]
```diff
@@ -192,38 +195,56 @@ impl WasiCtxBuilder {
...
+        inner.deterministic = Some(true);
+        rb_self
+    }
+
     pub fn build_context(&self, ruby: &Ruby) -> Result<wasmtime_wasi::WasiCtx, Error> {
-        let mut builder = wasmtime_wasi::WasiCtxBuilder::new();
         let inner = self.inner.borrow();
 
+        let mut wasi_ctx = if inner.deterministic.is_none() {
+            let mut ctx = wasmtime_wasi::WasiCtxBuilder::new();
```
Comment (cameronbarker) [inline ext/src/ruby_api/wasi_ctx_builder.rs:208]: I had to make the ctx mutable to allow overriding of `push_args` and `push_env`

File: spec/fixtures/deterministic.rs [hunk 5]
```diff
@@ -0,0 +1,85 @@
+//
```
Comment (cameronbarker) [inline spec/fixtures/deterministic.rs:1]: I included this file for:
1. Confirmation I'm writing code that would catch deterministic clock and random principals.
2. Ease of change for future updates.

**Review by saulecabrera — COMMENTED** [review]
Thinking a bit more about this, given that the `WasiCtxBuilder` and the deterministic version of it operate on different levels:
* The builder api operates on a `WasiCtxBuilder` and creates a `WasiCtx`
* The only way of creating a  `WasiCtx`, is through `WasiCtx::new` (which is what the deterministic crate uses).

I was thinking that it might not be totally crazy to introduce a proper `WasiCtx` class that's exposed to Ruby. This approach will reflect more accurately the reality of the Rust entities at play and will also make it easier to:

* Create a deterministic version
* Give Ruby users a way to interact with the `WasiCtx` class more freely (currently there's no way of doing this, the only entity that you can interact with is the `WasiCtxBuilder`)

Once we have this class exposed, we could have an API, that is something like:

```ruby
class WasiCtx
  def deterministic(...) # The deterministic version
  def set_* # All (or a subset of) the setters exposed by the Rust API
end
```

One of the disadvantages that I see with the current approach is that it mixes the `WasiCtx` and `WasiCtxBuilder` concepts, which might get tricky to manage if we ever end up adding more methods to the builder.

**Pull request closed without merging**